datafusion/core/src/execution/context.rs

```diff
@@ -676,6 +701,53 @@ impl SessionContext {
         Ok(false)
     }
 
+    /// Attempts to find a schema and deregister it. Returns a tuple of the schema and a
+    /// flag indicating whether dereg was performed (e.g if schema is found but has tables
+    /// then `cascade` must be set)
+    fn find_and_deregister_schema<'a>(
```

Comment:
I wonder if different catalogs might want to implement the "is the schema empty" check differently  🤔  

If so, it might make sense to put logic for "is the schema empty so can we drop it" in the `CatalogProvider`

Perhaps if `CatalogProvider` had a signature like:
```
    fn deregister_schema(&self, name: &str, cascade: bool) -> Result<Option<Arc<dyn SchemaProvider>>> {
```

I think this code would get significantly simpler as well as the `debug_assert`s to reassert what had just been validated would be unnecessary.

That being said, we could also do such a change as a follow on PR